File: 02/aoc_02.py

valid = 0 
valid2 = 0

def validate(char,string,upper,lower):

    cnt = 0
    for c in string:
        if c == char:
            cnt += 1
        
        if cnt > upper:
            return False
    
    return cnt >= lower
            
def validate_v2(char,string,u_ix,l_ix):

    if string[u_ix] == char:
        if string[l_ix] == char:
            return False
        return True
    
    if string[l_ix] == char:
        return True
    
    return False


with open('pass.txt','r') as f:
    ln = f.readline()
    while(ln):
        
        rg,c,p = ln.split()
        c = c[0]
        l,u = rg.split('-')
        
        if validate(c,p,int(u),int(l)):
            valid += 1

        if validate_v2(c,p,int(u)-1,int(l)-1):
            valid2 += 1

        ln = f.readline()
    
    print(f'Pt.1 {valid}')
    print(f'Pt.2 {valid2}')

File: 02/aoc_02.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func validatePart1(rng, char, pass string) bool {

	lStr := strings.Split(rng, "-")[0]
	uStr := strings.Split(rng, "-")[1]

	lower, _ := strconv.Atoi(lStr)
	upper, _ := strconv.Atoi(uStr)

	cnt := strings.Count(pass, char)

	return cnt >= lower && cnt <= upper
}

func validatePart2(rng, char, pass string) bool {

	lStr := strings.Split(rng, "-")[0]
	uStr := strings.Split(rng, "-")[1]

	lower, _ := strconv.Atoi(lStr)
	upper, _ := strconv.Atoi(uStr)

	if pass[lower-1] == char[0] {
		return pass[upper-1] != char[0]
	}

	return pass[upper-1] == char[0]

}

func main() {
	f, _ := os.Open("./pass.txt")
	defer func() {
		f.Close()
	}()

	validPart1 := 0
	validPart2 := 0
	s := bufio.NewScanner(f)
	for s.Scan() {
		parts := strings.Split(s.Text(), ":")

		pat := parts[0]
		pass := strings.TrimSpace(parts[1])
		patParts := strings.Split(pat, " ")

		if validatePart1(patParts[0], patParts[1], pass) {
			validPart1++
		}
		if validatePart2(patParts[0], patParts[1], pass) {
			validPart2++
		}
	}

	fmt.Printf("Ans part 1: %d\n", validPart1)
	fmt.Printf("Ans part 2: %d\n", validPart2)

}
